### ISSUE TITLE
`import/named` false positive when re-exporting internal dependencies with pnpm

### ISSUE THREAD

**Issue by teidesu** [import-js/eslint-plugin-import#2903]
When using PNPM, re-exporting internal dependencies triggers `import/named`, while TypeScript resolves everything correctly:
<img width="555" alt="image" src="https://github.com/import-js/eslint-plugin-import/assets/86301490/50568c2d-1d35-4ec1-9e0b-ae853a4f7afa">

# Steps to reproduce
```yaml
eslint: "8.52.0"
eslint-plugin-import: "2.28.1"
node: "18.12.1"
pnpm: "8.9.2"
```

I wasn't able to build a minimal repro in proper timeframe, but the general idea is that given

```typescript
// package-1/index.js
export * from 'package-2'

// package-2/index.js
export const a = 42
```

When installed via pnpm, results in roughly the following structure:
```
node_modules
├ .pnpm
│ ├ package-outer@1.0.0
│ │ └ node_modules
│ │   ├ package-inner -> .pnpm/package-inner@1.0.0/node_modules/package-inner
│ │   └ package-outer
│ │     └ ... source code ... 
│ └ package-inner@1.0.0
│   └ node_modules
│     └ package-inner
│       └ ... source code ...
└ package-outer -> .pnpm/package-outer@1.0.0/node_modules/package-outer
```

Where `eslint-plugin-import` likely fails to resolve `package-inner` when introspecting `package-outer`, which results in `import/named` false positives.

# Expected behaviour
Exports are correctly resolved from re-exported internal packages, no error is thrown

# Actual behaviour
`import/named` is thrown

# Possible workaround
The issue can be worked around by installing `package-outer` at the top level.

**Comment by dominic-p** [comment]
I'm seeing this as well. In my case, the re-export is from an internal file like this.

## library-package

**internal file: Button.jsx**

```js
export function CancelButton(props) {

    return (
        // ...
    );
}
```

**entry point: index.js**

```js
export * from './Button.jsx';
```

## consumer-package

```js
import { CancelButton } from 'library-package'
```

This gives linter error: CancelButton not found in 'library-package' eslint [import/named](https://github.com/import-js/eslint-plugin-import/blob/v2.32.0/docs/rules/named.md)

I'm not using pnpm or TypeScript in these projects (just npm and babel).

**Comment by ljharb** [comment]
@dominic-p which exact package is it? if it has `exports` and not `main`, we don't support that yet.

**Comment by ljharb** [comment]
and that index file, what module format is it using? If native ESM, i assume it has that named export? If transpiled ESM or CJS, what does module.exports look like?

**Comment by dominic-p** [comment]
I'm sorry, I'm not sure I'm following you. The index.js file in "library-package" looks just like I mentioned. It's just a series of reexports in native ESM format:

```js
export * from './Button.jsx';
export { default as AnotherComponent } from './AnotherComponent.jsx';
... 
```

The one thing that's a bit odd about my setup is that I'm distributing "library-package" as source. That is, I'm not bundling or transpiling it at all. So, the code above is exactly what you would see in "dist/index.js". Does that answer your question or am I still missing it?

**Comment by ljharb** [comment]
aha, yes that does :-) we don't yet support `exports` and we also don't yet support native `esm`. You'll need to use an alternative node resolver for the time being.

**Comment by dominic-p** [comment]
Ok, that makes sense. Sorry to hijack this issue then. :)
